### ISSUE TITLE
`swRrho()` and siblings cannot handle argo data

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#2208]
See below.  Here, `d` is an argo object.
```R
Rrho <- swRrho(d)
Error in gsw::gsw_Turner_Rsubrho(SA, CT, p) : 
  gsw_Turner_Rsubrho() cannot handle matrix SA
```

I grepped for all functions in `R/sw.R` and made the following checklist.  I'll pick away at them one by one, starting with things I use more than others.  

When I check something off, that means I have it working (*and* tested in `tests/testthat/test_sw.R` that the `[[X]]` results match the `swX()` results for both ctd and argo objects), but it does *not* mean that I have pushed to GH. In other words, this comment is a to-do list.

Items that are crossed out do not have `[[` methods.  I do not intend adding them for each case, at least at this time.  (Remember the "eat the dog food" principle.)

NOTE: the method I used for `swRrho()` -- recursively calling itself for columns of the matrix -- might be worth using elsewhere.  It's mainly needed for computations that are profile-based, another example of which is `swN2()` -- **which I will examine today or next week for possible rewriting**.

* [x] swAbsoluteSalinity
* ~swAlpha~  -- no [[ equivalent
* ~swAlphaOverBeta~ -- no [[ equivalent
* ~swBeta~ -- no [[ equivalent
* ~swCSTp~ -- no [[ equivalent
* [x] swConservativeTemperature
* [x] swDepth
* ~swDynamicHeight~ -- no [[ equivalent
* ~swLapseRate~ -- no [[ equivalent
* [x] swN2
* ~swPressure~ swPressure() computes pressure from z; it is not analogous with [["pressure"]]
* [x] swRho note: it is [["density"]].  I guess I could add a synonym, but who doesn't know what rho means?
* [x] swRrho
* ~swSCTp~ -- no [[ equivalent
* [x] swSigma
* [x] swSigma0
* [x] swSigma1
* [x] swSigma2
* [x] swSigma3
* [x] swSigma4
* [x] swSR
* ~swSTrho~ -- no [[ equivalent
* ~swSigmaT~ -- no [[ equivalent
* [x] swSigmaTheta
* ~swSoundAbsorption~ -- no [[ equivalent
* [x] swSoundSpeed
* ~swSpecificHeat~
* [x] swSpice
* ~swSpiciness~ -- no [[ equivalent
* [x] swSpiciness0
* [x] swSpiciness1
* [x] swSpiciness2
* [x] swSstar
* ~swTFreeze~ -- no [[ equivalent
* ~swTSrho~  -- no [[ equivalent
* ~swThermalConductivity~
* [x] swTheta
* ~swViscosity~
* [x] swZ

**Comment by dankelley** [comment]
I decided to make it a new issue.  I see how to fix it, and will do so tomorrow morning.

**Comment by dankelley** [comment]
Hang on.  The docs for `swRrho()` say it is for a `ctd` object.  So it should not even attempt to work with an `argo` object.  The code will get a bit tangled if I try to make it work for arbitrary objects.  

For example, consider the case where salinity and temperature are matrices.  If pressure is also a matrix, we can work by columns.  But maybe it's a vector, which applies across rows -- in that case, we have to check its type and see if it aligns properly, reporting an error if not.  That's 2 choices.  Then we have 2 choices for the EOS, so it's 4 choices.  And there are other possibilities.  The total number of choices is the product of all these choices, and we get into a classic case of code confusion.  All the pathways have to work and should (ideally) be tested in the test suite.

But there's another choice, which I think is much better: simply check the class and if it is `argo`, suggest the user try using `swRrho(as.ctd(argo, column=1))` etc.  A warning can be issued, and a note on this can be put in the docs.  That way, the old code won't need to be altered at its core, possibly changing behaviour in ways that might be surprising in some edge cases.

**Comment by dankelley** [comment]
This is now in the test suite (not pushed to GH ... `R/sw.R` is getting other changes for #2207 and I won't push until I have those things sorted -- perhaps today or tomorrow.  I'm only stating the test here as a note to myself. Note that the question of whether `as.ctd()` extracts profiles correctly is not within the domain of the present issue, but the whole point of writing *n* functions and testing them independently is that the aggregate test scales as *n*, not *n^2*.

```R
test_that("swRrho() signals error on argo objects (issue 2208)", {
    data(argo)
    expect_error(swRrho(argo), "cannot handle \"argo\" objects")
    e <- c(-1.4183550965, -1.3287788769, -1.0742467295, -0.7849644956, -0.4573938810, -0.1142189912)
    expect_equal(e, head(swRrho(as.ctd(argo, profile = 1)), 6))
})
```

**Comment by dankelley** [comment]
I grepped for all functions in `R/sw.R` and made the following checklist.  I'll pick away at them one by one, starting with things I use more than others.  When I check something off, that means I have it working (*and* tested in `tests/testthat/test_sw.R`), but it does *not* mean that I have pushed to GH. In other words, this comment is a to-do list.

* [ ] 282:swRrho
* [x] 448:swN2
* [ ] 565:swPressure
* [ ] 646:swCSTp
* [ ] 756:swSCTp
* [ ] 900:swSTrho
* [ ] 970:swTSrho
* [ ] 1092:swTFreeze
* [ ] 1175:swAlpha
* [ ] 1258:swAlphaOverBeta
* [ ] 1351:swBeta
* [ ] 1454:swThermalConductivity
* [ ] 1519:swDepth
* [ ] 1547:swZ
* [ ] 1648:swDynamicHeight
* [ ] 1765:swLapseRate
* [x] 1911:swRho
* [x] 1996:swSigma
* [ ] 2030:swSigmaT
* [x] 2099:swSigmaTheta
* [x] 2211:swSigma0
* [x] 2261:swSigma1
* [x] 2311:swSigma2
* [x] 2361:swSigma3
* [x] 2411:swSigma4
* [ ] 2500:swSoundAbsorption
* [ ] 2573:swSoundSpeed
* [ ] 2660:swSpecificHeat
* [ ] 2800:swSpice
* [ ] 2860:swSpiciness
* [x] 2927:swSpiciness0
* [x] 2962:swSpiciness1
* [x] 2996:swSpiciness2
* [ ] 3081:swTheta
* [ ] 3187:swViscosity
* [ ] 3249:swConservativeTemperature
* [ ] 3348:swAbsoluteSalinity
* [ ] 3449:swSstar
* [ ] 3541:swSR

**Comment by dankelley** [comment]
NOTE: I'm doing the recoding for this and 2207 in branch `issue_2207_2208`, not yet pushed to GH.  It may take a few more days to finish the work, and after that, I plan to use that branch for all my work for a few weeks.

**Comment by dankelley** [comment]
commit 2d47be5e8722db05a8fb827bf750de5c6639ec66
Author: dankelley <kelley.dan@gmail.com>
Date:   Sun Apr 28 20:07:50 2024 -0300

    improve (and test) many [[ methods for seawater computations
    
    This addresses issues #2207 and #2208.  See the checklist at the start
    of the former, for an overview.  For more details -- since I didn't keep
    the checklist up to date as I did the work -- look in the
    tests/testthat/test_sw.R and tests/testthat/test_accessors.R test files.
    
    I think all things are working now, and that I've trimmed the test
    printing code I had inserted.  But, especially to guard against the
    latter, I plan to use this branch in my daily work for a week or so,
    before merging into the "develop" branch.

**Comment by dankelley** [comment]
Oh, I see that the branch was not named in the previous comment.  It is "issues_2207_2208".

**Comment by dankelley** [comment]
I did an r-hub build, and all is OK except on the two test machines that cannot install the `ncdf4` package -- that is, the failure is nothing to do with `oce`.  I'm submitting this comment here and also on #2207.

**Comment by dankelley** [comment]
I think this is solved in "develop" commit fe7d581e0dc631a99b4ab3873919f15a440ca332
